pkg/graphviz/traceflow.go

```diff
@@ -0,0 +1,256 @@
+package graphviz
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/goccy/go-graphviz"
+	"github.com/goccy/go-graphviz/cgraph"
+	"k8s.io/klog"
+
+	opsv1alpha1 "github.com/vmware-tanzu/antrea/pkg/apis/ops/v1alpha1"
+)
+
+var (
+	clusterSrcName = "cluster_source"
+	clusterDstName = "cluster_dest"
+
+	// TODO to be updated
+	senderRole   = "sender"
+	receiverRole = "receiver"
+)
+
+func genSubGraph(graph *cgraph.Graph, result *opsv1alpha1.NodeResult, firstNodeName string, dir cgraph.DirType, addNodeNum int) []*cgraph.Node {
+	var nodes []*cgraph.Node
+	obs := result.Observations
+
+	// Show the name of cluster.
+	if len(result.Node) > 0 {
+		graph.SetLabel(result.Node)
+		if dir == cgraph.ForwardDir {
+			graph.SetLabelJust(cgraph.LeftJust)
+		} else {
+			graph.SetLabelJust(cgraph.RightJust)
+		}
+	}
+
+	// Construct the first node. Show it only if we know the name of it.
+	node, _ := graph.CreateNode(firstNodeName)
+	nodes = append(nodes, node)
+	if len(firstNodeName) > 0 {
+		node.SetFontColor("white")
+		node.SetColor("royalblue1")
+		node.SetStyle(cgraph.FilledNodeStyle)
+	} else {
+		node.SetStyle("invis")
+	}
+
+	// Reorder the observations according to the direction of edges.
+	if dir == cgraph.BackDir {
+		for i := len(obs)/2 - 1; i >= 0; i-- {
+			opp := len(obs) - 1 - i
+			obs[i], obs[opp] = obs[opp], obs[i]
+		}
+	}
+
+	// Draw the actual observations of traceflow.
+	for _, o := range obs {
+		// Construct node and edge.
+		nodeName := fmt.Sprintf("%s_%d", graph.Name(), len(nodes))
+		node, _ := graph.CreateNode(nodeName)
+		node.SetLabel(string(o.Component))
+		node.SetShape(cgraph.BoxShape)
+		node.SetStyle(cgraph.RoundedNodeStyle + "," + cgraph.FilledNodeStyle)
+		node.SetFontColor("white")
+		nodes = append(nodes, node)
+		if len(nodes) > 1 {
+			edgeName := fmt.Sprintf("%s_%d", graph.Name(), len(nodes))
+			edge, _ := graph.CreateEdge(edgeName, nodes[len(nodes)-2], nodes[len(nodes)-1])
+			edge.SetDir(dir)
+			edge.SetPenWidth(3.0)
+			edge.SetColor("limegreen")
+			if len(nodes) == 2 {
+				edge.SetMinLen(1 + addNodeNum)
+			} else {
+				edge.SetMinLen(1)
+			}
+			if o.Action == opsv1alpha1.Dropped && dir == cgraph.BackDir {
+				edge.SetStyle("invis")
+			}
+		}
+		// Set the pattern of node.
+		labelStr := string(o.Component) + "\n" + string(o.Action)
+		if o.Component == opsv1alpha1.NetworkPolicy && len(o.NetworkPolicy) > 0 {
+			labelStr += "\nNetpol: " + o.NetworkPolicy
+		}
+		if o.Action == opsv1alpha1.Dropped {
+			node.SetColor("violet")
+		} else {
+			node.SetColor("turquoise3")
+			if len(o.TunnelDstIP) > 0 {
+				labelStr += "\n" + o.TunnelDstIP
+			}
+		}
+		node.SetLabel(labelStr)
+	}
+	return nodes
+}
+
+// TODO question: is result and role 1-on-1 ?
+func getResultByRole(tf *opsv1alpha1.Traceflow, role string) *opsv1alpha1.NodeResult {
+	for _, result := range tf.Status.Results {
+		if result.Role == role {
```

Comment:
Role is not used yet, please check observations instead.
First observation of a Node:
```
	if isSender {
		ob := new(opsv1alpha1.Observation)
		ob.Component = opsv1alpha1.SpoofGuard
		ob.Action = opsv1alpha1.Forwarded
		obs = append(obs, *ob)
	} else {
		ob := new(opsv1alpha1.Observation)
		ob.Component = opsv1alpha1.Forwarding
		ob.Action = opsv1alpha1.Received
		ob.ComponentInfo = tableIDToComponentInfo[0]
		obs = append(obs, *ob)
	}
```